util/argo/ref_sources.go

```diff
@@ -0,0 +1,62 @@
+package argo
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	argoappv1 "github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1"
+	"github.com/argoproj/argo-cd/v2/util/db"
+)
+
+type GetRefSourcesOptions struct {
+	Spec       argoappv1.ApplicationSpec
+	Db         db.ArgoDB
+	Revisions  []string
+	IsRollback bool
+}
+
+// GetRefSources creates a map of ref keys (from the sources' 'ref' fields) to information about the referenced source.
+// This function also validates the references use allowed characters and does not define the same ref key more than
+// once (which would lead to ambiguous references).
+// In case of rollback, this function also updates the targetRevision to the proper revision
+func GetRefSources(ctx context.Context, opts GetRefSourcesOptions) (argoappv1.RefTargetRevisionMapping, error) {
+	refSources := make(argoappv1.RefTargetRevisionMapping)
+	if opts.Spec.HasMultipleSources() {
+		// Validate first to avoid unnecessary DB calls.
+		refKeys := make(map[string]bool)
+		for _, source := range opts.Spec.Sources {
+			if source.Ref != "" {
+				isValidRefKey := regexp.MustCompile(`^[a-zA-Z0-9_-]+$`).MatchString
+				if !isValidRefKey(source.Ref) {
+					return nil, fmt.Errorf("sources.ref %s cannot contain any special characters except '_' and '-'", source.Ref)
+				}
+				refKey := "$" + source.Ref
+				if _, ok := refKeys[refKey]; ok {
+					return nil, fmt.Errorf("invalid sources: multiple sources had the same `ref` key")
```

Comment:
it'd be really nice if errors like this reported at least a pair of sources so that users have something to chase

Reply:
I took the code from `argo.go` and I moved it as it is but I could add the key to the message if you think is useful

Reply:
My goal is to be able to read a message and figure out how to address it. If this error will be shown as is, I'd want to see the id of the thing (is it an `application/<x>`?) and two `sources` that had the same `ref` (preferably with the `ref`). If it's going to be pasted w/ other things, then as long as once it's pasted it has those bits, that'd make it usable.

Reply:
I'd try to improve the message 😄 
I'll review the logic trying to figure out the best way to print the message